File: intern/974.subarray-sums-divisible-by-k.js
/*
 * @lc app=leetcode id=974 lang=javascript
 *
 * [974] Subarray Sums Divisible by K
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const subarraysDivByK = function (nums, k) {
    const sumToCount = new Map();
    sumToCount.set(0, 1); // 和余数是 0 也符合题意

    let subarrayCount = 0;
    let subarraySumDivByK = 0;
    for (const num of nums) {
        subarraySumDivByK = (subarraySumDivByK + num) % k;
        if (subarraySumDivByK < 0) {
            // 只需要正余数
            subarraySumDivByK += k;
        }
        subarrayCount += sumToCount.get(subarraySumDivByK) ?? 0;
        sumToCount.set(
            subarraySumDivByK,
            (sumToCount.get(subarraySumDivByK) ?? 0) + 1,
        );
    }

    return subarrayCount;
};
// @lc code=end

File: src/2023-2024/974.subarray-sums-divisible-by-k.cpp
/*
 * @lc app=leetcode id=974 lang=cpp
 *
 * [974] Subarray Sums Divisible by K
 */

#include <memory>
#include <unordered_map>
#include <vector>

using std::unordered_map;
using std::vector;

// @lc code=start
class Solution {
 public:
  int subarraysDivByK(vector<int>& nums, int k) {
    const int N = nums.size();
    int prefixSumMod[2] = {0, 0};

    int* prefixSumModCount = new int[k];
    std::fill_n(prefixSumModCount, k, 0);
    // The prefix sum mod without all nums
    prefixSumModCount[0] = 1;
    int subArrayNumber = 0;

    for (int i = 0; i < N; i++) {
      const auto prevPrefixSumMod = i - 1 >= 0 ? prefixSumMod[(i - 1) % 2] : 0;
      prefixSumMod[i % 2] = ((prevPrefixSumMod + nums[i]) % k + k) % k;
      subArrayNumber += prefixSumModCount[prefixSumMod[i % 2]];
      prefixSumModCount[prefixSumMod[i % 2]]++;
    }

    delete[] prefixSumModCount;
    return subArrayNumber;
  }
};
// @lc code=end

int main() {
  vector<int> nums = {7, 4, -10};
  Solution sol;
  sol.subarraysDivByK(nums, 5);
}